File: leetCode-swift/leetCode-swift/leetCode-0006.swift
//
//  leetCode-0006.swift
//  leetCode-swift
//
//  Created by liyoucheng on 2018/11/21.
//  Copyright © 2018年 Giga. All rights reserved.
//

import Foundation



func convert(_ s: String, _ numRows: Int) -> String {
    if numRows <= 1 {
        return s
    }
    
    var res: [Character] = []
    let size = 2 * numRows - 2
    for i in 0..<numRows {
        for j in stride(from: i, to: s.count, by: size) {

            res.append(s[s.index(s.startIndex, offsetBy: j)])

            let temp = j + size - 2 * i
            if (i != 0 && i != numRows - 1 && temp < s.count) {
                
                res.append(s[s.index(s.startIndex, offsetBy: temp)])
            }
        }
    }
    
    return String(res)
}


func test_0006() {
    let s = convert("PAYPALISHIRING", 3)
    print(s)
}

File: leetCode-c/leetCode-c/LeetCode/leetCode-0006/leetCode-0006.c
//
//  leetCode-0006.c
//  leetCode-c
//
//  Created by liyoucheng on 2018/11/21.
//  Copyright © 2018年 Giga. All rights reserved.
//

#include "leetCode-0006.h"

#include <stdlib.h>
#include <string.h>

//http://www.cnblogs.com/springfor/p/3889414.html

char* convert(char* s, int numRows) {
    if (numRows <= 1) {
        return s;
    }
    
    int len = (int)strlen(s);
    char *new_str = malloc(len + 1);
    char *p = new_str;
    
    for (int row = 0; row < numRows; row++) {
        int interval1 = numRows + (numRows - 2) - row * 2;
        int interval2 = 2 * row;
        
        int i = row;
        int flag = 0;
        while (i < len) {
            *p++ = s[i];
            int delta = 0;
            do {
                delta = flag == 0 ? interval1 : interval2;
                flag = !flag;
            } while (delta == 0);
            i += delta;
        }
    }
    
    new_str[len] = '\0';
    return new_str;
}

void test_0006() {
    char *s = "PAYPALISHIRING";
    char *ret = convert(s, 3);
    printf("%s\n",ret);
}
